src/vcpkg/configuration.cpp

```diff
@@ -387,6 +382,81 @@ namespace
+    {
+        struct LocationAndRegistry
+        {
+            StringView location;
+            StringView registry;
+        };
+
+        // handle warnings from package pattern declarations
+        std::map<StringView, std::vector<LocationAndRegistry>> patterns;
+        for (auto&& reg : registries)
+        {
+            if (auto packages = reg.package_declarations.get())
+            {
+                for (auto&& pkg : *packages)
+                {
+                    auto it = patterns.find(pkg.pattern);
+                    if (it == patterns.end())
+                    {
+                        it = patterns.emplace(pkg.pattern, std::vector<LocationAndRegistry>{}).first;
+                    }
+                    it->second.emplace_back(LocationAndRegistry{
+                        pkg.location,
+                        reg.pretty_location(),
+                    });
+                }
+            }
+        }
+
+        auto append_declaration_warning = [](LocalizedString& msg,
+                                             const StringView& location,
+                                             const StringView& registry,
+                                             size_t indent_level) -> void {
+            msg.append_indent(indent_level)
+                .append(msgDuplicatePackagePatternLocation, msg::path = location)
+                .append_raw("\n")
+                .append_indent(indent_level)
+                .append(msgDuplicatePackagePatternRegistry, msg::url = registry)
+                .append_raw("\n");
+        };
+
+        std::vector<LocalizedString> warnings;
+        for (auto&& kvpair : patterns)
+        {
+            const auto& pattern = kvpair.first;
+            const auto& locations = kvpair.second;
+            if (locations.size() > 1)
+            {
+                auto first = locations.begin();
+                auto warning = msg::format_warning(msgDuplicatePackagePattern, msg::package_name = pattern)
+                                   .append_raw("\n")
+                                   .append_indent()
+                                   .append(msgDuplicatePackagePatternFirstOcurrence)
+                                   .append_raw("\n");
+                append_declaration_warning(warning, first->location, first->registry, 2);
+                warning.append_raw("\n")
+                    .append_indent()
+                    .append(msgDuplicatePackagePatternIgnoredLocations)
+                    .append_raw("\n");
+
+                auto last = locations.end();
+                for (auto cur = first + 1; cur != last; ++cur)
+                {
+                    append_declaration_warning(warning, cur->location, cur->registry, 2);
+                    if (cur + 1 != last)
+                    {
+                        warning.append_raw("\n");
+                    }
+                }
+                warnings.emplace_back(warning);
+            }
```

Comment:
```suggestion
            const auto& locations = kvpair.second;
            if (locations.size() > 1)
            {
                auto first = locations.begin();
                const auto last = locations.end();
                auto warning = msg::format_warning(msgDuplicatePackagePattern, msg::package_name = pattern)
                                   .append_raw("\n")
                                   .append_indent()
                                   .append(msgDuplicatePackagePatternFirstOcurrence)
                                   .append_raw("\n");
                append_declaration_warning(warning, first->location, first->registry, 2);
                warning.append_raw("\n")
                    .append_indent()
                    .append(msgDuplicatePackagePatternIgnoredLocations)
                    .append_raw("\n");
                    
                ++first;
                append_declaration_warning(warning, first->location, first->registry, 2);
                while (++first != last)
                {
                    warning.append_raw("\n");
                    append_declaration_warning(warning, first->location, first->registry, 2);
                }
                warnings.emplace_back(warning);
            }
```

Suggest eliminating the second iterator and avoiding the first/if antipattern by using the normal "join" pattern.